### PR TITLE
Add Reader classes

### LINKED ISSUE #1520: More flexible reader class with ability to introspect data file
Right now using `pyvista.read`, all data arrays are read and the first time point is used unless additional attributes are used. For example, this code will not read the 'temperature (C)' point data from the file saving memory.

```py
pyvista.read("probes.vtp", attrs={"SetPointArrayStatus": ['temperature (C)', 0]})
```

However, there is currently no way that I know of to inspect what data is available to read or what time points are available.

I'm proposing a class or classes that would work something like this:

```py
reader = pyvista.Reader("probes.vtp")
print(reader.point_array_names)
reader.disable_point_array('temperature (C)')
mesh = reader.read()
```

The vtk readers I use most often: vtkXMLReader, vtkOpenFOAMReader, and vtkGenericEnsightReader have the same user interface for enabling and disabling reading variables.  They also have similar ability to set the time point to read and introspection of the time.

I would envision a master class that defines the basic functionality and subclasses that extend functionality for more control.  This would be a parallel track for reading files that provides more control similar to `mesh.plot()` vs `pyvista.Plotter()`.

### PR DESCRIPTION
**Pull request by MatthewFlamm**
### Overview

This PR adds Reader classes that provides more fine-grained control of reading data files.  Some classes include the ability to inspect and enable/disable the reading of point and cell arrays.

closes #1520 

### Details

Future enhancements will include inspecting and setting time points to read, and more specific details for specific readers like vtkOpenFOAMReader, for example the ability to turn on and off the cell data to point data during data read.

The provided example shows the usage of the new class.

TODO:

- [x] Implement all XMLReaders
- [x] Implement openFOAM Reader
- [x] Implement GenericEnsightReader
- [x] Add list of supported file types
- [x] Tests  <- OpenFOAM tested locally and the ".p*" type files not tested
- [x] ~~Use observer functionality similar to `pyvista.read`~~ Not needed?

### CODE REVIEW

**Comment by akaszynski** [comment]
This is great!

**Comment by MatthewFlamm** [comment]
Did some more digging on the vtkDataSetReader, which is used for '.vtk' files.  [vtkDataReader docs](https://vtk.org/doc/nightly/html/classvtkDataReader.html) has the most useful methods.  It has a much different kind of structure than the ones listed in the TODO list.  Data is only differentiated by scalars vs. vector and as far as I can tell, you can only turn off/on all scalars or vectors, etc.

If the concept of scalars and vectors is added natively to pyvista as in #1431, then I'd propose we can, at that time, add this functionality into this class along with the vtkDataSetReader.

**Comment by MatthewFlamm** [comment]
Are there any openFOAM datasets we can use for testing?  I cannot find any.  Interestingly, the file in the backwards_facing_step download is named "foam_case_0_0_0_0.case" indicating it may have been generated using openFOAM???  @akaszynski uploaded this datafile, so you may be able to shed light on this.

**Comment by adeak** [comment]
The other issue linked in the linked issue has a link to a repo with a foam example https://github.com/pyvista/pyvista/issues/1511 -> https://github.com/fbob/PyVista_OpenFOAM_Tutorial. This was originally posted in slack.

**Comment by MatthewFlamm** [comment]
> The other issue linked in the linked issue has a link to a repo with a foam example #1511 -> https://github.com/fbob/PyVista_OpenFOAM_Tutorial. This was originally posted in slack, then I pointed this out in [#1510 (comment)](https://github.com/pyvista/pyvista/pull/1510#discussion_r667536179).

This data has no license associated with it, so I don't think we can use it?  Maybe @fbob can comment on its usability to be included in pyvista/vtk-data.  This would greatly help in adding openFOAM support to have a dataset to test against.

I could also cook up a quick example, but an already existing one would be preferable.

**Comment by MatthewFlamm** [comment]
I was going to add in observer directly into the Reader class, but now that we have a context manager, I think it is better to let the user manage it.  I've removed that from my TODO.  We could consider removing it from `pyvista.read` as well.

```py
with pyvista.VtkErrorCatcher() as error_catcher:
    reader = pyvista.Reader("some_file.vtp")
    mesh = reader.read()
```

**Comment by MatthewFlamm** [comment]
This PR is good to go except for testing the OpenFOAM Reader.  I have a proprietary dataset for which I tested manually and it works.  I'm going to mark this as "Ready for review".

I have another branch that adds in the OpenFOAM specific functionality of selecting mesh zones/patches and turning on and off the cell-to-point data translation.  We could consider dropping the OpenFOAM reader in this PR and then I can add the whole OpenFOAM Reader with all the bells and whistles later.

**Comment by MatthewFlamm** [comment]
One complication with this approach is a  potential ".pvd" reader.  See #414.  The array information only exists in the sub-files in ".pvd" format.  I was trying to keep this first PR smaller, but I wanted to raise some thornier implications downstream here.

An existing XMLReader could be nested inside to represent the active file about to be read:

```python
reader = Reader("file.pvd")
reader.set_time_value(1.0)
print(reader.active_file_reader.cell_arrays)
reader.read()  # equivalent to reader.active_file_reader.read()
```

But this approach only works if there is one file per time point.  If there are N files identified with different `part` numbers, we would have N different readers like:

```python
reader = Reader("file.pvd")
reader.set_time_value(1.0)
print(reader.active_file_readers[0].cell_arrays)
reader.read()  # equivalent to pyvista.MultiBlock([i_reader.read() for i_reader in reader.active_file_readers])
```

This would be similar to how you can select and deselect data inside paraview itself.  But in both of these cases, the `PVDReader` class would not be able to have the methods like `cell_arrays` defined directly, but these are defined in the `Reader` class in the PR as it stands.  We need a different class structure than the current PR, so this doesn't cause issues later.  I'll mark as Draft in the meantime.

**Comment by MatthewFlamm** [comment]
The class structure was reorganized so that different reader structures can be accomodated.  Now a Reader only has to implement the essential elements of a Reader.

1) returning the vtk reader
2) setting a filename
3) update
4) read

Now we can expose and set sane defaults for all the readers.  This PR shows an example of how this can be done for data handling that is shared by a group of readers using a Mixin class.  I added all the readers I could successfully test.  The readers for ".p" type vtk files like ".pvti" are included but aren't tested, as I assume these are similar in functionality to the base readers.

I haven't tackled the image file readers yet.  Maybe a second PR?

I eliminated the Reader class. It was a bit of magic, and now there is very little functionality implemented in the base class.  There is now a `get_reader` function which is more explicit in how it works to the user. This required renaming the same function in fileio.  If we prefer, I can easily go back to having a `Reader` class, so that the distinction between `pyvista.read` and the usage here is clearer.

**Comment by akaszynski** [comment]
> I eliminated the Reader class. It was a bit of magic, and now there is very little functionality implemented in the base class. There is now a get_reader function which is more explicit in how it works to the user.

I've used this approach before as well.  There are times where a "class factory" makes sense, but in this case (and as I've found in the past), the base class contains minimal functionality and it's better just to use a function instead.  In this case, I agree with using `get_reader`.

As for `read` vs `get_reader`, I don't think they'll be confused.

File: pyvista/utilities/reader.py
Comment (MatthewFlamm) [inline pyvista/utilities/reader.py:79]: I found that the OpenFOAM reader requires this to call `reader.reader.UpdateInformation()` to be able to set the time points correctly.  I need to test that this works for all the other readers before changing.

**Comment by MatthewFlamm** [comment]
The data from #1511 is from running the pitzDaily OpenFOAM tutorial.  I can regenerate it.  It isn't a time dependent data set _per se_, but it includes multiple sets of data at different iterations that show up as time dependent data to vtk.

File: pyvista/utilities/reader.py
Comment (MatthewFlamm) [reply pyvista/utilities/reader.py:79]: The example from #1511 was really helpful here.

**Comment by MatthewFlamm** [comment]
I was also having trouble thinking about how to add examples in the docstrings since the `download_*` functions already read the files. I'm thinking we should add an optional argument to allow for returning the file name instead:

```py
filename = download_bunny(filename=True)
reader = get_reader(filename)
```

**Comment by akaszynski** [comment]
> I was also having trouble thinking about how to add examples in the docstrings since the `download_*` functions already read the files. I'm thinking we should add an optional argument to allow for returning the file name instead:

We could also use a `load` parameter instead:
```py
filename = download_bunny(load=False)
reader = get_reader(filename)
```

I think this might make more sense, especially since many of our methods/functions use the `filename` parameter to specify the input of a filename.  Otherwise, we could use `as_filename` instead.

**Comment by MatthewFlamm** [comment]
> I think this might make more sense, especially since many of our methods/functions use the `filename` parameter to specify the input of a filename.

I agree, what about `return_filename`?  I think this is most explicit, but I don't have a strong feeling either way.

**Comment by akaszynski** [comment]
> I agree, what about `return_filename`? I think this is most explicit, but I don't have a strong feeling either way.

Perfect.  That's the most explicit.

**Comment by adeak** [comment]
To me,

  * `download_bunny(load=False)` means "_only_ return the path after download",
  * `download_bunny(return_filename=True)` means "_also_ return the filename after loading".

Different semantics.

**Comment by akaszynski** [comment]
> To me,
> 
> * `download_bunny(load=False)` means "_only_ return the path after download",
> * `download_bunny(return_filename=True)` means "_also_ return the filename after loading".
> 
> Different semantics.

That's a good point.  `numpy` does this with [numpy.unique](https://numpy.org/doc/stable/reference/generated/numpy.unique.html).  Enabling the `return` parameters returns additional variables.  Perhaps `load` is best in this case.

**Comment by MatthewFlamm** [comment]
All great points.  I'll go with `load`.  I don't see a use case for returning the loaded mesh and the filename.

**Comment by MatthewFlamm** [comment]
> The data from #1511 is from running the pitzDaily OpenFOAM tutorial. I can regenerate it.

I looked into this and was able to regenerate the data. However, since this is from the OpenFOAM tutorial and the case setup is part of the results, I'm unsure if the OpenFOAM license would apply, i.e. GPL v3.  If we cannot use this data, I could try to cook up my own case, but it would take longer.  I could also delete the case setup from the results, but there are some options in the vtkOpenFOAMReader that utilize the case setup.  It would be nice to have these files for testing this functionality.
 I'm not 100% sure that this case can utilize those options, so it may be a moot point.

**Comment by akaszynski** [comment]
> I could also delete the case setup from the results, but there are some options in the vtkOpenFOAMReader that utilize the case setup. It would be nice to have these files for testing this functionality.
I'm not 100% sure that this case can utilize those options, so it may be a moot point.

Avoiding any GPL v3 issues is a priority for me and this project as it's used in a variety of commercial and non-commercial libraries.  Reduced testing coverage is a small price to pay to avoid being copyleft.

**Comment by MatthewFlamm** [comment]
> Avoiding any GPL v3 issues is a priority for me and this project as it's used in a variety of commercial and non-commercial libraries. Reduced testing coverage is a small price to pay to avoid being copyleft.

I'm glad I raised it.  I will continue testing on my closed source datasets periodically, but we won't get any OpenFOAM tests until I, or someone else, can cook up a more open dataset.

**Comment by MatthewFlamm** [comment]
One topic for discussion is that currently the readers do not inherit directly from the vtk reader classes, although the DataSet classes do.  The upside to the approach here is that we better control the interface presented to the user.  The downside is that we cannot use the pyvista Reader instances in a vtk workflow, but one can still use `reader.reader` to accomplish this.  This is my preferred style as it limits the clutter of methods presented to the user, but I'm open to changing it especially since it is inconsistent with other classes.

**Comment by akaszynski** [comment]
> The downside is that we cannot use the pyvista Reader instances in a vtk workflow, but one can still use `reader.reader` to accomplish this.

This works, we should provide documentation for this in the `reader` property showing an example use case to plug into a vtk pipeline.

**Comment by akaszynski** [comment]
@MatthewFlamm, is this ready for review?

**Comment by MatthewFlamm** [comment]
It is ready for review.  I'm only working on getting the documentation included after #1588 was merged.

File: pyvista/utilities/reader.py
Comment (MatthewFlamm) [inline pyvista/utilities/reader.py:350]: I originally had this as 

```py
>>> sliced_mesh.plot(scalars="Void Volume Fraction",  # Doc building complained of EOL here
...                  cpos='xz', show_scalar_bar=False)

```

This passed doctest locally, but failed in building the documentation.  It looks a little long to me on one line, but I couldn't figure out how to break itup.

Comment (adeak) [reply pyvista/utilities/reader.py:350]: This sounds really weird. We have a bunch of these elsewhere in the docs. E.g. https://github.com/pyvista/pyvista/blob/fab7f777aa4af5d9007cef12c77738236e66e77e/pyvista/core/filters/data_set.py#L2842-L2844 rendered at https://dev.pyvista.org/api/core/_autosummary/_autosummary/pyvista.DataSetFilters.streamlines_evenly_spaced_2D.html#pyvista.DataSetFilters.streamlines_evenly_spaced_2D.

Are you sure there wasn't some other typo there? Missing comma or something like that?

Comment (MatthewFlamm) [reply pyvista/utilities/reader.py]: :shrug:  I pushed the version that fails in the documentation but passes in the doctest as it will probably be more helpful for debugging.

Maybe it is the way I run doctest???  I am using `pytest --doctest-modules pyvista/utilities/reader.py`

Comment (MatthewFlamm) [reply pyvista/utilities/reader.py]: It passed here, but fails on my local doc build.  Curious.  Maybe it is a dependency version thing?

**Comment by MatthewFlamm** [comment]
I will need some guidance on the documentation as I have to stumble wildly through sphinx and rst.  I copied and pasted my way to get a working state for one group of Reader Classes, the XML RectilinearGrid Readers.  Since all the methods and attributes are part of the classes that are inherited, `BaseReader` and `PointCellDataSelection`, there is really only an example shown .  Should we include all inherited methods/attributes for the readers, using `:inherited-members:`, or is the link to the super classes enough?

image

**Comment by akaszynski** [comment]
> I will need some guidance on the documentation as I have to stumble wildly through sphinx and rst. I copied and pasted my way to get a working state for one group of Reader Classes, the XML RectilinearGrid Readers. Since all the methods and attributes are part of the classes that are inherited, `BaseReader` and `PointCellDataSelection`, there is really only an example shown . Should we include all inherited methods/attributes for the readers, using `:inherited-members:`, or is the link to the super classes enough?

Let me take a look at this later today and evaluate if we should include inherited or not.

**Review by akaszynski — APPROVED** [review]
I've reworked the documentation a bit to follow our template pattern:
tmp


I prefer having all inherited methods explicitly shown rather than telling the user to navigate through additional layers of documentation ("Flat is better than nested")

Other than the changes to the docs, I'm happy with this PR.  Approved and nice work!  Merge when ready.

**Comment by MatthewFlamm** [comment]
This looks nice!

There are a few things that I'd like to clean up before merging:  1) Duplicate MultiBlock reference in the toctree and 2) to add the `get_reader` function somewhere

**Comment by MatthewFlamm** [comment]
I tried doing this, but my ability to build the documentation seems to have broken again.

I get a ton of "autosummary: stub file not found" errors even when building from your last commit in this PR, which seems weird to me.  As you can see, none of the pages get linked correctly here.

image

**Comment by akaszynski** [comment]
> I tried doing this, but my ability to build the documentation seems to have broken again.

Sometimes I have to run `make -C doc clean` to get the docs to build correctly.  Let me check this locally.

**Comment by akaszynski** [comment]
Build looks great and links work.  Made one final change:
sc0
sc1

**Pull request merged**
